### ISSUE TITLE
persistedsqlstats: timeout during TestSQLStatsFlush under stressrace

### ISSUE THREAD

**Issue by michae2** [cockroachdb/cockroach#115744]
Hit a timeout during an upgrade while running TestInMemoryStatsDiscard under stressrace.

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/13000619?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
=== RUN   TestInMemoryStatsDiscard
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestInMemoryStatsDiscard194623428
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:159: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
panic: test timed out after 5m24s
```

Artifacts are attached.
[Tests_maybe_stressrace_161361_artifacts.zip](https://github.com/cockroachdb/cockroach/files/13596068/Tests_maybe_stressrace_161361_artifacts.zip)

Jira issue: CRDB-34179

**Comment by michae2** [comment]
I was not able to reproduce this with `./dev test pkg/sql/sqlstats/persistedsqlstats --filter=TestInMemoryStatsDiscard --stress --race --count=85000 --ignore-cache` which ran overnight on my gceworker.

**Comment by stevendanna** [comment]
I wonder if the issue here was actually with the other tests in this package. The entire test timed out after 5m24s, but from this output it looks like TestInMemoryStatsDiscard had only been running for 15s while the proceeding two tests were running for most of the 5m:


```
--- PASS: TestPersistedSQLStatsReset (101.41s)
... SNIP ...
--- PASS: TestSQLStatsFlush (207.09s)
=== RUN   TestInMemoryStatsDiscard
... SNIP ..
panic: test timed out after 5m24s
running tests:
	TestInMemoryStatsDiscard (15s)
```

**Comment by michae2** [comment]
> I wonder if the issue here was actually with the other tests in this package. The entire test timed out after 5m24s, but from this output it looks like TestInMemoryStatsDiscard had only been running for 15s while the proceeding two tests were running for most of the 5m:

Ah, good catch. I got [another timeout with TestSQLStatsFlush](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/13131175?buildTab=overview&expandBuildProblemsSection=true&hideProblemsFromDependencies=false&expandBuildTestsSection=true&hideTestsFromDependencies=false&expandBuildChangesSection=true), so that does seem like the culprit.

Aha, this is a dupe of https://github.com/cockroachdb/cockroach/issues/116003.

Thanks for looking!

**Comment by yuzefovich** [comment]
fixed by backports of #116073
